src/main/scala/is/hail/expr/ir/BinaryOp.scala

```diff
@@ -34,11 +35,36 @@ object BinaryOp {
   private def incompatible[T](lt: Type, rt: Type, op: BinaryOp): T =
     throw new RuntimeException(s"Cannot apply $op to $lt and $rt")
 
-  def emit(op: BinaryOp, lt: Type, rt: Type, l: Code[_], r: Code[_]): Code[_] = (lt, rt) match {
+  def emit(fb: FunctionBuilder[_], mb: StagedBitSet, op: BinaryOp, lt: Type, rt: Type, lm: Code[Boolean], lv: Code[_], rm: Code[Boolean], rv: Code[_]): (Code[Unit], Code[Boolean], Code[_]) = {
+    emitPropagatesNA(op, lt, rt, lv, rv).map {
+      (Code._empty, lm || rm, _)
+    }
+      .orElse({
+        val xlm = mb.newBit()
```

Comment:
I think this will be more sustainable as a separate method, that way we'd have `emitPropagatesNA` which is NA-ignorant as well as this method which is NA-aware.